### ISSUE TITLE
Latest PHP 5.4, 5.5 & 5.6 versions are not being used per the PHP build definitions

### ISSUE THREAD

**Issue by ntwb** [travis-ci/travis-ci#7014]
Per https://github.com/php-build/php-build/tree/master/share/php-build/definitions
And https://secure.php.net/downloads.php

PHP `5.6.x`, the latest is `5.6.29` yet `5.6.5` is used: https://travis-ci.org/aaronjorbin/develop.wordpress/jobs/183138649#L165

PHP `5.5.x`, the latest is `5.5.38` yet `5.5.21` is used: https://travis-ci.org/aaronjorbin/develop.wordpress/jobs/183138647#L161

PHP `5.4.x`, the latest is `5.4.45` yet `5.4.37` is used: https://travis-ci.org/aaronjorbin/develop.wordpress/jobs/183138646#L161

PHP `7.0.x`, `7.1.x`, `5.3.x` and `5.2.x` branches are unaffected and using the latest branch versions

`.travis.yml`. source: https://github.com/aaronjorbin/develop.wordpress/blob/master/.travis.yml

**Comment by BanzaiMan** [comment]
Thank you for taking time to open this issue.

`M.N` means "the latest version of M.N" pre-installed on the build image. The teeny version will depend on the cookbooks configuration at the time build images are provisioned.

Unfortunately, this sort of discrepancy is inevitable; PHP release cycles are fairly steady while our build image updates do not. There will be times when the latest PHP release is not pre-installed on our images.

We do provide on-demand installations of those runtimes that are not found. This is how 7.0, 7.1, 5.3, and 5.2 *appear* to work as you expect. "5.{4,5,6}" are pre-installed, and the pre-installed versions are used to save time on builds.

If you wish to use more recent teeny version releases, please specify the `x` value.

This is also documented in https://docs.travis-ci.com/user/languages/php#Choosing-PHP-versions-to-test-against:

> which are specified using aliases for the “most recent x.y.z release” provided on Travis CI of any given line. For a full listing of the supported versions see About Travis CI Environment.

I am closing this now, but if you have further questions, do let us know.

**Comment by BanzaiMan** [comment]
Well, 5.2 and 5.3 are actually slightly different. They appear to work because they are no longer maintained and the "latest" version is pre-installed (on Precise images, that is).

**Comment by ntwb** [comment]
Thanks @BanzaiMan, always gleam lots of insight into the magic of Travis CI from your replies 👍 

That said, I do have a couple of follow ups if you don't mind 😄 

> Unfortunately, this sort of discrepancy is inevitable; PHP release cycles are fairly steady while our build image updates do not. There will be times when the latest PHP release is not pre-installed on our images.

Using PHP 5.6 as an example, PHP 5.6.5 was added 23 January 2015 via https://github.com/php-build/php-build/commit/2718f0d60dc9bae86fbd0575e3a7202b4f8fbc5e

And PHP 5.6.29 was added ~12 hours ago via https://github.com/php-build/php-build/commit/2080889de66008c0103c687dadd6afc2a705d14a#diff-8621f5423e123b35f811fed649e47903

So, I would have thought at least one newer version than PHP 5.6.5 would have been added to the Travis CI pre-built images in the past ~23 months

Can and where would I submit a PR for the  `5.4`, `5.5`, and `5.6` aliases be updated to a somewhat newer version?

Also highlighted in the linked docs is:
```
Specifying exact versions like 5.3.8 is discouraged as it may break your build when we update PHP versions on Travis CI.
```

This is why in the past I've not used `x.y.z`, I've presumed `phpenv` would act similar to `nvm` for NodeJS versions and the `x.y` would always map (and install) if needed the latest `x.y.z` version from the `x.y` branch.

**Comment by BanzaiMan** [comment]
I'd be happy to answer them, of course!

The addition of the build definition in php-build/php-build unfortunately does not trigger the packaging process. This step is still manual; we try to be as responsive as we can, but sometimes we can lag behind. You can see when things happen in https://travis-ci.org/travis-ci/php-src-builder/builds (or you can ping us). Incidentally, we triggered a build for 5.6.29 a few hours ago: https://travis-ci.org/travis-ci/php-src-builder/builds/183122855 The build image updates after Feb 2015 were particularly slow to come, for many reasons (mostly to make our infrastructure more stable and easier to manage). For this, we apologize.

It is funny that you mention the passage in the PHP documentation. I was just reading that last Friday, and thinking that we need some clarification around that.

The comparison with `nvm` is natural, but unfit here, I must say. Node.js runtimes are very quick to install via binary archive (via `nvm`), while PHP runtimes require certain configuration and compilation.

I hope this answers your questions.

**Comment by BanzaiMan** [comment]
Well, the last point about PHP runtime installation is not quite accurate.

It *is* possible to write the build script in such a way that `x.y` always points to the latest version of `x.y.z` available, very much like `nvm` does. This adds some complexity to the build script, which I would personally like to avoid. There is also a question of user expectations; I am not sure what the _right_ thing is.

**Comment by ntwb** [comment]
Right, thanks again, yeah, I've followed along the infrastructure updates, Docker, GCE etc so I'm rather familiar with that, also rather familiar with the PHP 5.2 issues on the newer infrastructure also, none of it is easy for sure 😏 

As to the _right_ thing to do, I'm not sure of this either, I like the idea of `x.y.z` always pointing to the latest version, but as stated this is not as simple as NodeJS for PHP. Maybe clearing up the documentation somewhat would be sufficient.

Now looking more at the `php-src-builder` builds:
• https://travis-ci.org/travis-ci/php-src-builder/builds/174809244: `VERSION=5.6.28 ALIAS=5.6`
• https://travis-ci.org/travis-ci/php-src-builder/builds/176523428: `VERSION=5.6.27 ALIAS=5.6`

> "Incidentally, we triggered a build for 5.6.29 a few hours ago: https://travis-ci.org/travis-ci/php-src-builder/builds/183122855"

Going by the description of each of those that I linked above were they not the same case where the PHP alias was being updated to `5.6.29`, `5.6.28`, and `5.6.27` respectively?

And if that is the case this job submitted ~30 minutes ago still uses `5.6.5` yet the `5.6` alias was updated to `5.6.29` a _few hours ago /shrug

**Comment by BanzaiMan** [comment]
Aliases `5.6`, `5.5`, etc., are useful when we provision build images, at which point the aliases are set up for use. Once the images are deployed, the aliases will point to the pre-installed versions, and specifying `5.5` will find the version that is pre-installed on that image.

**Comment by jaymecd** [comment]
It would be awesome to upgrade alias of `7.0`, eg, `7.0.14`, as `phpdbg` on `7.0.7` has some issues with coverage generation for phpunit

**Comment by BanzaiMan** [comment]
@jaymecd Sorry, but that will have to wait until the images are updated.

**Comment by jaymecd** [comment]
@BanzaiMan when it could happen, approx?

**Comment by sebastianbergmann** [comment]
What is the point of aliases such as `5.6` when they do not give you the latest version of that version series?

The current situation where `5.6` gives you PHP 5.6.5 (without PHPDBG) is a major impediment to the development of PHPUnit, for instance, as the build for one of its core dependencies always fails: https://travis-ci.org/sebastianbergmann/php-code-coverage/jobs/204494698.

**Comment by TysonAndre** [comment]
https://docs.travis-ci.com/user/build-environment-updates/2017-09-06/ - It seems like they'll add 5.6.31 a week from now (September 6th)
https://blog.travis-ci.com/2017-08-29-trusty-image-updates

**Comment by BanzaiMan** [comment]
@sebastianbergmann 

> What is the point of aliases such as 5.6 when they do not give you the latest version of that version series?

The point is to provide a reasonably recent 5.6.x release without demanding the users to update their configuration every time there is a new release or a new build image.

We aim to minimize the delta by having frequent build image updates.

Thanks for your understanding.

**Comment by edmorley** [comment]
If it helps, I've bumped the versions of PHP pre-installed into the images to the latest patch releases in travis-ci/packer-templates#548. This is available via `group: edge` now, and will be on `group: stable` next time `edge` is promoted to stable.
